### PR TITLE
Create default categories and add them to current users

### PR DESCRIPTION
**Pull request by marderer**
- Create default categories for new users
- Create tasks for creating default categories for current users. Just run `rails user:generate_categories`

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -2,6 +2,7 @@ class User < ApplicationRecord
   has_many :transactions, dependent: :destroy
   has_many :authorizations, dependent: :destroy
   has_many :categories, dependent: :destroy
+  after_create :create_default_categories
```
Comment (AleksSenkou) [inline app/models/user.rb]: @marderer let's not use `after_...` and `before_...` hooks for models, because of implicity. See [post](https://medium.com/planet-arkency/the-biggest-rails-code-smell-you-should-avoid-to-keep-your-app-healthy-a61fd75ab2d3)

Instead of hooks we can add `Users::Creator` service for creating users: it will be useful because of two reasons:
1) all stuff that we need for user creating is in one place
2) we can reuse this service, even from a console, for example

Comment (AleksSenkou) [reply app/models/user.rb]: @marderer so, here is what I see as a fast solution:
```ruby
class Users::Creator < Struct.new(:email)
  def call
    create_user
    add_default_categories
  end

  private

  def create_user
    @user = User.create(
      email: email,
      password: generate_password,
      confirmed_at: Date.current
    )
  end

  def generate_password
    Devise.friendly_token[0, 20]
  end

  def add_default_categories
    Categories::Creator.new(@user).create_default_categories
  end
end
```
and in `Users::Omniauth`:
```ruby
def user
  @_user ||= User.find_by(email: email) || create_user
end

def create_user
  Users::Creator.new(email).call
end
```

File: app/models/user.rb [hunk 2]
```diff
@@ -13,4 +14,8 @@ class User < ApplicationRecord
   def balance
     Users::Balance.new(self).balance
   end
+
+  def create_default_categories
+    CreateCategotyService.new(self).create_default_categories
```
Comment (AleksSenkou) [inline app/models/user.rb]: @marderer what do you think about `Categories::Creator`? sounds easier

File: lib/tasks/user.rake [hunk 3]
```diff
@@ -0,0 +1,8 @@
+namespace :user do
+  desc 'Creating default categories for all current users'
+  task generate_categories: :environment do
+    User.all.each do |user|
+      CreateCategotyService.new(user).create_default_categories
+    end
+  end
```
Comment (AleksSenkou) [inline lib/tasks/user.rake]: @marderer 👍 
don't forget about `find_each`

File: app/models/user.rb (hunk 1 above)
Comment (marderer) [reply app/models/user.rb]: @AleksSenkou it's fine, but service for creating a user returns category list. I added `@user` to `#add_default_category` method. Is it good or better add `return @user` to #call method?

**Comment by paneq** [comment]
You could also publish and event that a User was created http://railseventstore.org/docs/publish/ and have a handler subscribed http://railseventstore.org/docs/subscribe/ which would create default categories.

**Comment by AleksSenkou** [comment]
@paneq that's a great thing to have, thank you. I like `RailsEventStore` - it helped me with collecting metrics.
The current project is a test one, so, time is limited. We'll think about adding event sourcing to the next one ;)

File: lib/tasks/user.rake [hunk 4]
```diff
@@ -0,0 +1,8 @@
+namespace :user do
+  desc 'Creating default categories for all current users'
+  task generate_categories: :environment do
```
Comment (AleksSenkou) [inline lib/tasks/user.rake]: @marderer what do you think about renaming it the task to `generate_default_categories`? Cause it's the thing tasks do

File: app/services/users/creator.rb [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    )
+  end
+
+  def generate_password
+    Devise.friendly_token[0, 20]
+  end
+
+  def add_default_categories
+    Users::DefaultCategoriesCreator.new(@user).call
+    @user
```
Comment (AleksSenkou) [inline app/services/users/creator.rb]: let's return `@user` from `#call` instead of `#add_default_categories`. `add_default_categories` doesn't say anything about a user, you can't expect that it returns `@user`

File: app/services/users/default_categories_creator.rb [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+    default_categories_list.map { |title| create_category(title) }
+  end
+
+  private
+
+  def create_category(title)
+    user_categories.create(title: title, amount: 0, user_id: user.id)
+  end
+
+  delegate :categories, to: :user, prefix: true
```
Comment (AleksSenkou) [inline app/services/users/default_categories_creator.rb]: @marderer why did you add delegate here and not at the end of the file? Can you try that?

File: app/services/users/default_categories_creator.rb [hunk 7]
```diff
@@ -0,0 +1,17 @@
+class Users::DefaultCategoriesCreator < Struct.new(:user)
+  def call
+    default_categories_list.map { |title| create_category(title) }
+  end
+
+  private
+
+  def create_category(title)
```
Comment (AleksSenkou) [inline app/services/users/default_categories_creator.rb]: @marderer private methods with arguments say that a service is overloaded and you have to pass arguments to private methods

what do you think about:
```ruby
def call
  default_categories_list.map do |name|
    user_categories.create(name: name, amount: 0)
  end
end
```
?

File: app/services/users/default_categories_creator.rb [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+    user_categories.create(title: title, amount: 0, user_id: user.id)
```
Comment (AleksSenkou) [inline app/services/users/default_categories_creator.rb]: @marderer you don't have to pass `user_id` because you are creating categories from a user association

File: app/services/users/creator.rb (hunk 5 above)
Comment (marderer) [reply app/services/users/creator.rb]: @AleksSenkou Oh, sorry, I repair it later, forgot repair here.

File: app/services/users/default_categories_creator.rb (hunk 6 above)
Comment (marderer) [reply app/services/users/default_categories_creator.rb]: @AleksSenkou here too

File: app/services/users/default_categories_creator.rb (hunk 8 above)
Comment (marderer) [reply app/services/users/default_categories_creator.rb]: @AleksSenkou and here :(

File: app/services/users/default_categories_creator.rb (hunk 7 above)
Comment (marderer) [reply app/services/users/default_categories_creator.rb]: @AleksSenkou Maybe better this method make private? And call in the method `call`? 

``` ruby
  def call
    create_default_categories
  end

  private

  def create_default_categories
    default_categories_list.map do |name|
      user_categories.create(name: name, amount: 0)
    end
  end

```

Comment (AleksSenkou) [reply app/services/users/default_categories_creator.rb]: @marderer hah, here is a play with names, when you call just one method in `call` says that we need just one method called `#create_default_categories` instead of `#call` 😄 

So, we can go with one public method `#create_default_categories`

File: app/models/category.rb [hunk 9]
```diff
@@ -2,6 +2,7 @@ class Category < ApplicationRecord
   belongs_to :user
 
   validates :amount, :title, presence: true
-  validates :amount, numericality: {greater_than_or_equal_to: 0}
+  validates :amount, presence: true, format: {with: /\A\d+(?:\.\d{0,2})?\z/},
+                     numericality: {greater_than_or_equal_to: 0, less_than: 1_000_000_00}
```
Comment (AleksSenkou) [inline app/models/category.rb]: `100_000_000`

Comment (marderer) [reply app/models/category.rb]: @AleksSenkou I looked [here](http://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Style/NumericLiterals) and used syntax for cents.
`1_000_000_00` means `1 000 000.00` not 100 000 000

File: db/migrate/20180323134330_chenges_type_of_amount_in_categories_and_transactions.rb [hunk 10]
```diff
@@ -0,0 +1,7 @@
+class ChengesTypeOfAmountInCategoriesAndTransactions < ActiveRecord::Migration[5.1]
+  def change
+    remove_column :categories, :amount, :string
+    add_column :categories, :amount, :decimal, precision: 8, scale: 2
+    change_column :transactions, :amount, :decimal, precision: 8, scale: 2
```
[the remark this answers is not in the document]
Comment (marderer) [reply db/migrate/20180323134330_chenges_type_of_amount_in_categories_and_transactions.rb]: @AleksSenkou yes, I tried so do, but was problems. I thought that here I could use  recreate after removing, because all categories have `amount: 0`.

[the remark this answers is not in the document]
Comment (AleksSenkou) [reply db/migrate/20180323134330_chenges_type_of_amount_in_categories_and_transactions.rb]: @marderer let's keep safe policy while data migration
